File: server/haskell/src/Record/Create.hs
module Record.Create (create) where

import Auth.JWT (authorizedUserId)
import Control.Monad (when)
import Control.Monad.IO.Class (liftIO)
import Data.Aeson.TH (defaultOptions, deriveJSON)
import Data.Int (Int64)
import Data.Maybe (fromMaybe)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import Data.Time (UTCTime)
import qualified Data.Time as Clock
import qualified Hasql.Pool as HP
import qualified Hasql.Session as HS
import qualified Hasql.Statement as S
import qualified Hasql.TH as TH
import Network.HTTP.Types (status201, status400, status500)
import Web.Scotty
  ( ActionM,
    ScottyM,
    json,
    jsonData,
    liftAndCatchIO,
    post,
    raiseStatus,
    status,
  )

data CreateRequest = CreateRequest
  { board_size :: Int,
    name :: Maybe String,
    black_player :: Maybe String,
    white_player :: Maybe String,
    comment :: Maybe String,
    handicap :: Maybe Int,
    komi :: Maybe Double,
    ruleset :: Maybe String
  }
  deriving (Eq, Show)

$(deriveJSON defaultOptions ''CreateRequest)

data CreateResponse = CreateResponse
  { id :: Int,
    owner :: Int,
    board_size :: Int,
    created :: UTCTime,
    name :: String,
    black_player :: String,
    white_player :: String,
    comment :: String,
    handicap :: Int,
    komi :: Double,
    ruleset :: String,
    winner :: String
  }
  deriving (Eq, Show)

$(deriveJSON defaultOptions ''CreateResponse)

type RecordRow = (Int64, Int64, T.Text, T.Text, T.Text, T.Text, Int64, Double, T.Text, T.Text, UTCTime)

insert :: S.Statement RecordRow Int64
insert =
  [TH.singletonStatement|
    insert into record_record
    (owner_id, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created)
    values
    ($1 :: int8, $2 :: int8, $3 :: text, $4 :: text, $5 :: text, $6 :: text, $7 :: int8, $8 :: float8, $9 :: text, $10 :: text, $11 :: timestamptz)
    returning id :: int8
  |]

withDefaults :: Int64 -> UTCTime -> CreateRequest -> RecordRow
withDefaults userId now request = do
  ( userId,
    fromIntegral board_size,
    T.pack $ fromMaybe "" name,
    T.pack $ fromMaybe "Black" black_player,
    T.pack $ fromMaybe "White" white_player,
    T.pack $ fromMaybe "" comment,
    fromIntegral $ fromMaybe 0 handicap,
    fromMaybe 7.5 komi,
    T.pack $ fromMaybe "AGA" ruleset,
    "U",
    now
    )
  where
    CreateRequest {board_size, name, black_player, white_player, comment, handicap, komi, ruleset} = request

toResponse :: Int64 -> RecordRow -> CreateResponse
toResponse recordId (userId, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created) =
  CreateResponse
    { id = fromIntegral recordId,
      owner = fromIntegral userId,
      board_size = fromIntegral board_size,
      created,
      name = T.unpack name,
      black_player = T.unpack black_player,
      white_player = T.unpack white_player,
      comment = T.unpack comment,
      handicap = fromIntegral handicap,
      komi,
      ruleset = T.unpack ruleset,
      winner = T.unpack winner
    }

create :: HP.Pool -> ScottyM ()
create pool = post "/api/records/" $ do
  userId <- authorizedUserId
  request <- jsonData :: ActionM CreateRequest
  let CreateRequest {board_size} = request
  when (board_size /= 9 && board_size /= 13 && board_size /= 19) $ raiseStatus status400 "Invalid board size"
  now <- liftAndCatchIO Clock.getCurrentTime
  let row = withDefaults userId now request
  result <- liftIO $ HP.use pool $ HS.statement row insert
  case result of
    Right recordId -> do
      status status201
      json $ toResponse recordId row
    Left err -> do
      raiseStatus status500 $ TL.pack $ show err

File: server/rust/src/record/create.rs
use axum::{extract::State, http::StatusCode, response::IntoResponse, Json};
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio_postgres::Client;

use crate::auth::UserId;
use crate::db::query_one;

#[derive(Serialize, Deserialize)]
struct CreateRequest {
    board_size: u32,
    name: Option<String>,
    black_player: Option<String>,
    white_player: Option<String>,
    comment: Option<String>,
    handicap: Option<i32>,
    komi: Option<f64>,
    ruleset: Option<String>,
}

#[derive(Serialize, Deserialize)]
struct CreateResponse {
    id: i64,
    owner: i32,
    board_size: u32,
    created: DateTime<Utc>,
    name: String,
    black_player: String,
    white_player: String,
    comment: String,
    handicap: i32,
    komi: f64,
    ruleset: String,
    winner: String,
}

pub async fn create(
    State(client): State<Arc<Client>>,
    UserId(user_id): UserId,
    body: String,
) -> impl IntoResponse {
    let CreateRequest {
        board_size,
        name,
        black_player,
        white_player,
        comment,
        handicap,
        komi,
        ruleset,
    } = serde_json::from_str(&body).unwrap();
    if ![9, 13, 19].contains(&board_size) {
        return Err((StatusCode::BAD_REQUEST, "Invalid board size"));
    }
    let name = name.unwrap_or("".to_string());
    let black_player = black_player.unwrap_or("Black".to_string());
    let white_player = white_player.unwrap_or("White".to_string());
    let comment = comment.unwrap_or("".to_string());
    let handicap = handicap.unwrap_or(0);
    let komi = komi.unwrap_or(7.5);
    let ruleset = ruleset.unwrap_or("AGA".to_string());
    let winner = "U".to_string();
    let now: chrono::DateTime<chrono::Utc> = chrono::Utc::now();
    let row = query_one(&client, "INSERT INTO record_record (owner_id, board_size, name, black_player, white_player, comment, handicap, komi, ruleset, winner, created) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11) RETURNING id", &[&user_id, &TryInto::<i32>::try_into(board_size).unwrap(), &name, &black_player, &white_player, &comment, &handicap, &komi, &ruleset, &winner, &now]).await?;
    let record_id = row.get("id");
    Ok((
        StatusCode::CREATED,
        Json(CreateResponse {
            id: record_id,
            owner: user_id,
            board_size,
            created: now,
            name,
            black_player,
            white_player,
            comment,
            handicap,
            komi,
            ruleset,
            winner,
        }),
    ))
}
